text/3869-ignore-derive.md

```diff
@@ -0,0 +1,622 @@
+[summary]: #summary
+
+The `#[ignore]` attribute can now be applied to fields.
+Its purpose is to tell derive macros to skip the field when generating code.
+
+```rust
+#[derive(Clone, PartialEq, Eq, std::hash::Hash)]
+struct User {
+    #[ignore(PartialEq, std::hash::Hash)]
+    //       ^^^^^^^^^  ^^^^^^^^^^^^^^^
+    //       traits that will ignore this field
+    name: String,
+    #[ignore(PartialEq, std::hash::Hash)]
+    age: u8,
+    id: u64
+}
+```
+
+For the above struct `User`, derives `PartialEq` and `Hash` will ignore the `name` and `age` fileds.
+Code like this is generated:
+
+```rust
+impl Clone for User {
+    fn clone(&self) -> User {
+        User {
+            name: self.name.clone(),
+            age: self.age.clone(),
+            id: self.id.clone(),
+        }
+    }
+}
+
+impl PartialEq for User {
+    fn eq(&self, other: &User) -> bool {
+        self.id == other.id
+    }
+}
+
+impl Eq for User {}
+
+impl std::hash::Hash for User {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) -> () {
+        std::hash::Hash::hash(&self.id, state)
+    }
+}
+```
+
+# Motivation
+[motivation]: #motivation
+
+It's common to want to exclude one or more fields when deriving traits such as `Debug` or `Deserialize`.
+To do this, you currently need to completely abandon the `derive` and instead implement the trait manually.
+
+Manually implementing the trait is much more error-prone than letting the derive do it for you.
+For example, when a new field is added, it's possible to forget to change all of your implementations.
+
+This is why it's idiomatic to instead `derive` traits when it's possible to do so. But this deriving isn't possible if you need to ignore a field.
+Common use-cases of ignoring fields when implementing traits include:
+
+- Types that are high-level wrappers around other types, which may include some additional metadata but ultimately delegate most trait
+  implementations to the inner type. An example is an identifier in programming languages:
+
+  ```rust
+  /// Identifier for a variable, function or trait
+  #[derive(PartialEq, Eq)]
+  struct Ident {
+      /// Location of the identifier in the source code
+      #[ignore(PartialEq, Eq)]
+      span: Span,
+      /// Value of the identifier
+      value: String
+  }
+  ```
+  
+- Minor improvements such as skipping irrelevant fields in a `Debug` derive (like fields of type `PhantomData`) derive will become
+  easy enough for people to just do. Currently, the effort required to maintain a manual implementation is too high -
+  so people just don't bother with it in most cases.
+
+- Security: Data with sensitive fields like AWS s3 `access_key_id` and `secret_access_key` that **must** be skipped for `Debug` implementations is forced to be implemented manually - leading to a significant increase in boilerplate
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When using `#[derive]`, you can apply `#[ignore]` to fields:
```

Comment:
it could also be useful to `#[ignore]` enum variants for things like `Deserialize`